### PR TITLE
[SPARK-28554][SQL] JDBC v2 with catalog functionalities

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR implements `JDBCTableCatalog`, so that end-users can use the JDBC as a catalog.

### Why are the changes needed?

JDBC is a good fit for the new catalog plugin API. It's good to have one builtin implementation of the catalog plugin API.

### Does this PR introduce any user-facing change?

Yes, now end-users can register JDBC as a catalog.

### How was this patch tested?

a new test suite.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rdblue @brkyvz @gatorsmile

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -133,7 +133,7 @@ statement
         createFileFormat |
         locationSpec |
         (TBLPROPERTIES tableProps=tablePropertyList))*                 #createTableLike
-    | replaceTableHeader ('(' colTypeList ')')? tableProvider
+    | replaceTableHeader ('(' colTypeList ')')? tableProvider?
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: TODO: move it to a separate PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -731,12 +731,11 @@ class Analyzer(
     extends Rule[LogicalPlan] with LookupCatalog {
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case s @ ShowTables(UnresolvedNamespace(Seq()), _) =>
-        s.copy(namespace =
-          ResolvedNamespace(currentCatalog.asNamespaceCatalog, catalogManager.currentNamespace))
+        s.copy(namespace = ResolvedNamespace(currentCatalog, catalogManager.currentNamespace))
       case UnresolvedNamespace(Seq()) =>
-        ResolvedNamespace(currentCatalog.asNamespaceCatalog, Seq.empty[String])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We don't load the namespace metadata here and the catalog doesn't need to be a namespace catalog. For example, SHOW TABLE is implemented by `TableCatalog.listTables`, and is nothing to do with `SupportsNamespace`.

TODO: move it to a separate PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTable.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  private var isTruncate = false
+
+  override def truncate(): WriteBuilder = {
+    isTruncate = true
+    this
+  }
+
+  override def buildForV1Write(): InsertableRelation = new InsertableRelation {
+    override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+      // TODO: do truncate and append atomically.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTable.scala]: If you don't mind, please file a JIRA issue and make this IDed TODO like `TODO(SPARK-XXX)`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      properties: util.Map[String, String]): Table = {
+    checkNamespace(ident.namespace())
+    if (partitions.nonEmpty) {
+      throw new UnsupportedOperationException("Cannot create JDBC table with partition")
+    }
+    // TODO: we can support this, but we need to add an API to `JdbcDialect` to generate the SQL
+    //       statement to specify table options.
+    if (!properties.isEmpty) {
+      logWarning("Cannot create JDBC table with properties, these properties will be " +
+        "ignored: " + properties.asScala.map { case (k, v) => s"$k=$v" }.mkString("[", ", ", "]"))
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala]: For security, shall we redaction the value? Or, it will be enough to print only `key` list?
cc @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  import testImplicits._
+
+  val tempDir = Utils.createTempDir()
+  val url = s"jdbc:h2:${tempDir.getCanonicalPath};user=testUser;password=testPass"
+  var conn: java.sql.Connection = null
+
+  override def sparkConf: SparkConf = super.sparkConf
+    .set("spark.sql.catalog.h2", classOf[JDBCTableCatalog].getName)
+    .set("spark.sql.catalog.h2.url", url)
+    .set("spark.sql.catalog.h2.driver", "org.h2.Driver")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala]: Could we have a test case that have two JDBC catalogs?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      val rs = md.getTables(null, schemaPattern, "%", Array("TABLE"));
+      val tables = ArrayBuffer.empty[Identifier]
+      while (rs.next()) {
+        tables += Identifier.of(namespace, rs.getString("TABLE_NAME"))
+      }
+      tables.toArray
+    }
+  }
+
+  override def tableExists(ident: Identifier): Boolean = {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala]: The implementation of `tableExists` is different with `JdbcUtils.tableExists`. Should we unify them?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    // TODO: support this by adding more APIs to `JdbcDialect` which can generate the SQL statement
+    //       to alter a table.
+    throw new UnsupportedOperationException("alter table")
+  }
+
+  override def dropTable(ident: Identifier): Boolean = {
+    checkNamespace(ident.namespace())
+    try {
+      withConnection { conn =>
+        val statement = conn.createStatement
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala]: This is duplicated with `JdbcUtils.dropTable`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTable.scala [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  override def newWriteBuilder(info: LogicalWriteInfo): WriteBuilder = {
+    val mergedOptions = new JdbcOptionsInWrite(
+      jdbcOptions.parameters.originalMap ++ info.options.asCaseSensitiveMap().asScala)
+    new JDBCWrite(schema, mergedOptions)
+  }
+}
+
+class JDBCScan(relation: JDBCRelation) extends ScanBuilder with V1Scan
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTable.scala]: Nit: how about moving `JDBCScan` to a single file?

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTable.scala]: Also, it would be great if we split it into JDBCScanBuilder and JDBCScan, for demonstrating the design idea of DSV2 API.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalog.scala]: @dongjoon-hyun good point, +1

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+import org.apache.spark.sql.{QueryTest, Row}
+import org.apache.spark.sql.catalyst.analysis.CannotReplaceMissingTableException
+import org.apache.spark.sql.catalyst.plans.logical.Filter
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+import org.apache.spark.sql.execution.datasources.v2.jdbc.JDBCTableCatalog
+import org.apache.spark.sql.functions.lit
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.util.Utils
+
+class JDBCV2Suite extends QueryTest with SharedSparkSession {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala]: Should we reuse part of the test cases in `JDBCSuite` here?
